imagetagger/imagetagger/images/views.py

```diff
@@ -189,111 +192,108 @@ def upload_image(request, imageset_id):
-                                img_fname = (''.join(shortname) + '_' +
-                                             ''.join(
-                                                 random.choice(
-                                                     string.ascii_uppercase + string.ascii_lowercase + string.digits)
-                                                 for _ in range(6)) + extension)
-                                try:
-                                    with PIL_Image.open(file_path) as image:
-                                        width, height = image.size
-                                    file_new_path = os.path.join(imageset.root_path(), img_fname)
-                                    shutil.move(file_path, file_new_path)
-                                    shutil.chown(file_new_path, group=settings.UPLOAD_FS_GROUP)
-                                    new_image = Image(name=filename,
-                                                      image_set=imageset,
-                                                      filename=img_fname,
-                                                      checksum=fchecksum,
-                                                      width=width,
-                                                      height=height
-                                                      )
-                                    new_image.save()
-                                except (OSError, IOError):
-                                    error['damaged'] = True
-                                    os.remove(file_path)
+                                image_file.seek(0)
+                                fchecksum = fchecksum.digest()
+                                # Tests for duplicates in imageset
+                                if Image.objects.filter(checksum=fchecksum,
+                                                        image_set=imageset).count() == 0:
+                                    (shortname, extension) = splitext(filename)
+                                    img_fname = (''.join(shortname) + '_' +
+                                                 ''.join(
+                                                     random.choice(
+                                                         string.ascii_uppercase + string.ascii_lowercase + string.digits)
+                                                     for _ in range(6)) + extension)
+                                    try:
+                                        with PIL_Image.open(image_file) as image:
+                                            width, height = image.size
+                                        image_file.seek(0)
+                                        imageset_dir.upload(img_fname, image_file)
+                                        # TODO rfrigg: Check if necessary
+                                        # shutil.chown(file_new_path, group=settings.UPLOAD_FS_GROUP)
```

Comment:
@Akasch Can you explain why this is needed? I guess it might be needed for an OS filesystem, but not for a cloud filesystem like GCS. I can look into adding some code that adds this line back in case an OS filesystem is used.

Reply:
@ftsell Shall we remove these TODOs since you tested everything and it works as expected?

Reply:
Yes although I'm not 100% sure either, I think it is safe to do so.